### PR TITLE
feat: Initialise application controller with chart pull & push functionality

### PR DESCRIPTION
**Pull request by kushthedude**
Signed-off-by: Kush Trivedi <kushthedude@gmail.com>

### CODE REVIEW

File: config/crd/bases/orkestra.azure.microsoft.com_applicationgroups.yaml [hunk 1]
```diff
@@ -4,7 +4,7 @@ apiVersion: apiextensions.k8s.io/v1beta1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.2.9
+    controller-gen.kubebuilder.io/version: v0.2.5
```
Comment (nitishm) [inline config/crd/bases/orkestra.azure.microsoft.com_applicationgroups.yaml:7]: Can you upgrade your kubebuilder binary to v0.2.9

File: controllers/application_controller.go [hunk 2]
```diff
@@ -30,12 +31,32 @@ type ApplicationReconciler struct {
...
+	}
+
+	if application.Status.ChartStatus.Ready {
+		return ctrl.Result{}, nil
+	}
+
+	requeue, err = r.reconcile(logr, &application)
+	if err != nil {
+		logr.Error(err, "failed to reconcile application instance")
+		return ctrl.Result{}, err
```
Comment (nitishm) [inline controllers/application_controller.go:56]: `ctrl.Result{Requeue: requeue}` here as well, as an error might be benign at times and we might still want to requeue

**Review by nitishm — CHANGES_REQUESTED** [review]
Looks good so far. Just address the comments

File: controllers/application_controller.go [hunk 3]
```diff
@@ -30,12 +31,32 @@ type ApplicationReconciler struct {
 // +kubebuilder:rbac:groups=orkestra.azure.microsoft.com,resources=applications/status,verbs=get;update;patch
```
Comment (nitishm) [inline controllers/application_controller.go:32]: Can you add the `updateStatusAndEvent()` as well to the reconcile function

File: config/crd/bases/orkestra.azure.microsoft.com_applicationgroups.yaml (hunk 1 above)
Comment (kushthedude) [reply config/crd/bases/orkestra.azure.microsoft.com_applicationgroups.yaml]: done

File: controllers/application_controller.go (hunk 3 above)
Comment (kushthedude) [reply controllers/application_controller.go:32]: done

File: api/v1alpha1/application_types.go [hunk 4]
```diff
@@ -16,6 +16,7 @@ type ApplicationSpec struct {
 	// Namespace to which the HelmRelease object will be deployed
 	Namespace                string `json:"namespace"`
 	Subcharts                []DAG  `json:"subcharts,omitempty"`
+	GroupName                string `json:"groupname"`
```
Comment (nitishm) [inline api/v1alpha1/application_types.go:19]: The json tag is required to be `groupName` by kubebuilder (controller-runtime).

Comment (nitishm) [reply api/v1alpha1/application_types.go:19]: Let's call this GroupID

File: pkg/registry/fetch.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		// todo(kushthedude): do we want to create a config
+	}
+	var err error
+	if location == "" {
+		homeDir, err := os.UserHomeDir()
+		if err != nil {
+			logr.Error(err, "unable to get home directory")
+			return "", fmt.Errorf("unable to get home directory: %w", err)
+		}
+		destFolder := "orkestra/charts"
```
Comment (nitishm) [inline pkg/registry/fetch.go:37]: move this into a const

File: pkg/registry/fetch.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	return chartLocation, err
+}
+
+// Debug used to give output in a structured manner for all helm related registry operation
+func Debug(format string, v ...interface{}) {
+	format = fmt.Sprintf("[debug] %s\n", format)
+	_ = log.Output(2, fmt.Sprintf(format, v...))
+}
```
Comment (nitishm) [inline pkg/registry/fetch.go:78]: do we need this right now? We should converge on a logging standard if we do need this.

File: pkg/registry/fetch.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	helmopv1 "github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	"github.com/go-logr/logr"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/cli"
+	"log"
+	"os"
+	"path"
+)
+
+func Fetch(helmReleaseSpec helmopv1.HelmReleaseSpec, location string, logr logr.Logger) (string, error) {
```
Comment (nitishm) [inline pkg/registry/fetch.go:14]:
```suggestion
func Fetch(logr logr.Logger, helmReleaseSpec helmopv1.HelmReleaseSpec, location string) (string, error) {
```
More idiomatic to pass supporting variables like config, context or logger as the first set of args.

**Review by nitishm — CHANGES_REQUESTED** [review]

**Comment by nitishm** [comment]
I am wondering if much of the helm CLI init can be abstracted away.
The registry package functions leverage the same logger that is passed through the reconciler. Additionally for every call to the registry functions the client is initialized,
```go
	actionCfg := new(action.Configuration)

	settings := cli.New()
	helmDriver := "memory"
	if err := actionCfg.Init(settings.RESTClientGetter(), settings.Namespace(), helmDriver, Debug); err != nil {
		logr.Error(err, "unable to initialize action configuration")
		return "", fmt.Errorf("unable to initialize action configuration: %w", err)
	}

	var chartLocation string
	pullInstance := action.NewPull()
	pullInstance.Settings = settings
	 ...
```

Instead, we could abstract it into a registry struct and set all the shared variables once. Something along the lines of 
```go
type Client struct {
    logr logr.Logger
    cfg *action.Config
    ...
}
```
That we instantiate the registry.Client() once during the controller initialization in `main.go` and then reuse that instant across the reconcilers.

File: api/v1alpha1/application_types.go [hunk 8]
```diff
@@ -16,13 +16,20 @@ type ApplicationSpec struct {
 	// Namespace to which the HelmRelease object will be deployed
 	Namespace                string `json:"namespace"`
 	Subcharts                []DAG  `json:"subcharts,omitempty"`
+	GroupId                  string `json:"groupId"`
```
Comment (nitishm) [inline api/v1alpha1/application_types.go:19]:
```suggestion
	GroupID                  string `json:"groupID"`
```

**Review by nitishm — APPROVED** [review]
Let's push this to a new branch. Call it `dev` and I will push my PR for ApplicationGroup into that as well

**Comment by kushthedude** [comment]
> Let's push this to a new branch. Call it `dev` and I will push my PR for ApplicationGroup into that as well

Sounds great, will merge into development once I fix golangci-lint 😄

**Pull request merged**